src/ray/core_worker/reference_count.cc

```diff
@@ -157,13 +157,19 @@ void ReferenceCounter::AddOwnedObject(const ObjectID &object_id,
   // If the entry doesn't exist, we initialize the direct reference count to zero
   // because this corresponds to a submitted task whose return ObjectID will be created
   // in the frontend language, incrementing the reference count.
-  object_id_refs_.emplace(object_id, Reference(owner_address, call_site, object_size,
-                                               is_reconstructable, pinned_at_raylet_id));
+  auto it = object_id_refs_
+                .emplace(object_id, Reference(owner_address, call_site, object_size,
+                                              is_reconstructable, pinned_at_raylet_id))
+                .first;
   if (!inner_ids.empty()) {
     // Mark that this object ID contains other inner IDs. Then, we will not GC
     // the inner objects until the outer object ID goes out of scope.
     AddNestedObjectIdsInternal(object_id, inner_ids, rpc_address_);
   }
+  if (pinned_at_raylet_id.has_value()) {
+    // We eagerly add the pinned location to the set of object locations.
+    AddObjectLocationInternal(it, pinned_at_raylet_id.value());
```

Comment:
Is this because the normal RPC to add the object location is async?

I think it makes sense to add it preemptively, but I'm a bit worried about the potential of a memory leak here. Are we certain that this entry will get deleted? Considering message reordering, failures, etc.

Actually is message reordering not an issue for the normal Add/Remove RPCs?

Reply:
Another option is to open this part in a separate PR, since it doesn't look like it's needed for the main change of using the cached locations.

Reply:
>Is this because the normal RPC to add the object location is async?

Yes, the motivation is a combination of (1) reporting the pinned location as early as possible for the sake of future scheduling decisions that might otherwise miss out on that information, and (2) simplifying the `ReferenceCounter::GetLocalityData` logic by ensuring that `ref.pinned_at_raylet_id` is always in `ref.locations`. Before this change, I had a lot of logic around only using `ref.locations` if the object was owned by us, OBOD is turned on, _and_ `!ref.locations.empty()`, otherwise use `ref.pinned_at_raylet_id`, but only if it has a value, etc. Adding the invariant that `ref.locations` always contains the full object locations, as up-to-date as possible, even if the OBOD is turned off (in which case it will only contain the `ref.pinned_at_raylet_id` location), greatly simplified the logic there and elsewhere. A side-benefit is that the OBOD will now have even more up-to-date information.

>I think it makes sense to add it preemptively, but I'm a bit worried about the potential of a memory leak here. Are we certain that this entry will get deleted? Considering message reordering, failures, etc. Actually is message reordering not an issue for the normal Add/Remove RPCs?

We currently make some very strong assumptions about message ordering and network reliability: if the add/remove messages for a given location are reordered or if the remove message never reaches the owner, the location will be added and never removed until the reference is deleted. It was my impression, back when we decided that a failed OBOD --> core worker request is equivalent to the object being freed, that we were punting on these kind of concerns for now. Btw, if anything, adding the `pinned_at_raylet_id` location earlier lessens the chance that message reordering would result in a leak!

Also, do we currently handle message reordering for the GCS-based object directory? I thought that we don't.

>Another option is to open this part in a separate PR, since it doesn't look like it's needed for the main change of using the cached locations.

I was going to, but it does make the `ReferenceCounter::GetLocalityData` logic much, much cleaner, so I somewhat consider it in scope? 😁  If I pulled it out into a different PR, I'd have to reimplement that core `GetLocalityData` logic twice instead of once, in this PR.

Reply:
I see, okay it makes sense. We should definitely take another look about message ordering, though. The grpc bindings don't guarantee message ordering between a client and server. Can you open an issue for this so we can track it? It would be best to try and find a stress test that uncovers this scenario.